### PR TITLE
Avoid torch amp cuda warning with bf16 on cpu

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Use precision_plugin context manager to avoid following warning with bf16 on CPU.

```console
UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
```

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

Part of #1

cc @borda @carmocca @justusschock @awaelchli @akihironitta

### CODE REVIEW

File: pytorch_lightning/utilities/model_summary.py [hunk 1]
```diff
@@ -261,16 +261,10 @@ def _forward_example_input(self) -> None:
...
-            and trainer.amp_backend == AMPType.NATIVE
-            and trainer._device_type != _AcceleratorType.TPU
-        ):
-            model.forward = torch.cuda.amp.autocast()(model.forward)
-
         mode = model.training
         model.eval()
-        with torch.no_grad():
+
+        with torch.no_grad(), (trainer.precision_plugin.forward_context() if trainer else contextlib.nullcontext()):
```
Comment (justusschock) [inline pytorch_lightning/utilities/model_summary.py:267]: I think this syntax won't work with older python versions (3.6 and 3.7)

File: CHANGELOG.md [hunk 2]
```diff
@@ -322,6 +322,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed double evaluation bug with fault-tolerance enabled where the second call was completely skipped ([#11119](https://github.com/PyTorchLightning/pytorch-lightning/pull/11119))
 
+
+- Avoided torch amp cuda warning with `bf16` on cpu ([#11161](https://github.com/PyTorchLightning/pytorch-lightning/pull/11161))
```
Comment (justusschock) [inline CHANGELOG.md:326]:
```suggestion
- Avoided torch amp cuda warning with `bf16` on cpu when summarizing ([#11161](https://github.com/PyTorchLightning/pytorch-lightning/pull/11161))
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/model_summary.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/model_summary.py:267]: yeah looks like just py3.6.
Will wait for this: https://github.com/PyTorchLightning/pytorch-lightning/pull/11117

File: pytorch_lightning/utilities/model_summary.py [hunk 3]
```diff
@@ -261,16 +261,10 @@ def _forward_example_input(self) -> None:
         input_ = model.example_input_array
         input_ = model._apply_batch_transfer_handler(input_)
 
-        if (
-            trainer is not None
-            and trainer.amp_backend == AMPType.NATIVE
-            and trainer._device_type != _AcceleratorType.TPU
-        ):
-            model.forward = torch.cuda.amp.autocast()(model.forward)
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_summary.py:269]: is this a bug in the current code that we patch the `forward` without ever resetting it?

Comment (rohitgr7) [reply pytorch_lightning/utilities/model_summary.py:269]: maybe not since eventually all the steps will be happening on precision=16

File: pytorch_lightning/utilities/model_summary.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/model_summary.py]: done!

File: CHANGELOG.md [hunk 4]
```diff
@@ -322,6 +322,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed double evaluation bug with fault-tolerance enabled where the second call was completely skipped ([#11119](https://github.com/PyTorchLightning/pytorch-lightning/pull/11119))
 
+
+- Avoided torch amp cuda warning with `bf16` on cpu when summarizing ([#11161](https://github.com/PyTorchLightning/pytorch-lightning/pull/11161))
```
Comment (awaelchli) [inline CHANGELOG.md:326]:
```suggestion
- Fixed an incorrect warning being produced by the model summary when using `bf16` precision on CPU ([#11161](https://github.com/PyTorchLightning/pytorch-lightning/pull/11161))
```

**Review by awaelchli — APPROVED** [review]
nice to see this being generalized.

**Review by ananthsub — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
